src/v/io/page.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <seastar/core/temporary_buffer.hh>
+
+#include <cstdint>
+
+namespace experimental::io {
+
+/**
+ * A page represents a contiguous region of data in a file.
+ */
+class page {
+public:
+    /**
+     * Construct a page with the given \p offset and \p data.
+     */
+    page(uint64_t offset, seastar::temporary_buffer<char> data);
+
+    page(const page&) = delete;
+    page& operator=(const page&) = delete;
+    page(page&&) = delete;
+    page& operator=(page&&) = delete;
+    ~page() = default;
+
+    /**
+     * Offset of this page in the underlying file.
```

Comment:
nit: maybe it'll be more obvious when we use this more in practice, but wdyt about calling this `filepos` or `position` or something? Just thinking it might be good to disambiguate these from model::offset or kafka::offsets early

Reply:
I agree, but I think it might be more important to ensure that this byte-oriented offset never leaks out of the `io` subsystem, which it really shouldn't. Of course offset/filepos/etc will still be part of an abstraction at the `io` subsystem boundary, but those will largely be part of the standarized seastar::input_stream/output_stream types.

Reply:
Yeah agreed it's for sure more important to ensure correctness at that boundary. The name change would be more to help readers maintain a clear mental separation when jumping between subsystems (though maybe that won't happen much in practice? not sure)

Reply:
>though maybe that won't happen much in practice?

That's what I think. I'm hesitant to use `filepos` (or some other name) to describe an `offset` because what happens when we have another situation where we have an _offset_ -- are we are going to run out of names? Maybe I should just create a strong/named_type to use in the io subysystem?

Reply:
>what happens when we have another situation where we have an offset

I'm sure there'd always be some alternative, but feel free to punt if you think it won't be a concern. I'm just a little wary of already overloaded terms, particularly when they're already ingrained in the adjacent codebase

>Maybe I should just create a strong/named_type to use in the io subysystem?

+1 yeah i think that'd help (at least it appeases my concern)

Reply:
ok cool will do